python/ray/autoscaler/_private/resource_demand_scheduler.py

```diff
@@ -39,6 +50,56 @@
 # The minimum number of nodes to launch concurrently.
 UPSCALING_INITIAL_NUM_NODES = 5
 
+NodeResources = ResourceDict
+ResourceDemands = List[ResourceDict]
+
+
+class UtilizationScore(Protocol):
+    """This fancy class just defines the `UtilizationScore` protocol to be
+    some type that is a "totally ordered set" (i.e. things that can be sorted).
```

Comment:
Actually, I guess no one's actually enforcing transitivity of the relation, so it's not literally a totally ordered set, but w/e.

Reply:
> Actually, I guess no one's actually enforcing transitivity of the relation

Well python doesn't enforce anything, but I think transitivity is required? Yeah I guess in (type) theory we should formally express contravariance (you can't compare objects that are _different_ subtypes of UtilizationScore).